File: apps/posting/posting_core/metrics/maintenance.py
from __future__ import annotations

import os
import sqlite3
from contextlib import contextmanager
from datetime import datetime, timezone, timedelta
from pathlib import Path

from posting_core.metrics_config import SCHEDULED_TARGETS, iso_utc, utc_now_dt

UNSUPPORTED_METRIC_TARGETS = ("x", "linkedin")
MAINTENANCE_LOCK_NAME = "metrics_maintenance"


def split_targets(value: str | None) -> tuple[str, ...]:
    if not value:
        return tuple(SCHEDULED_TARGETS)
    requested = tuple(part.strip() for part in value.split(",") if part.strip())
    return tuple(target for target in requested if target in SCHEDULED_TARGETS)


def build_backfill_plan(conn, targets: tuple[str, ...], refs: list[str] | None = None, date_from: str | None = None, date_to: str | None = None):
    params: list[object] = []
    where = ["p.status='active'", "t.status='published'"]
    placeholders = ",".join("?" for _ in targets)
    where.append(f"t.target IN ({placeholders})")
    params.extend(targets)
    if refs:
        ref_placeholders = ",".join("?" for _ in refs)
        where.append(f"p.post_key IN ({ref_placeholders})")
        params.extend(refs)
    if date_from:
        where.append("p.date_utc >= ?")
        params.append(date_from)
    if date_to:
        where.append("p.date_utc <= ?")
        params.append(date_to)
    rows = conn.execute(
        f"""
        SELECT p.post_key, p.post_id, p.message_id, p.date_utc, t.target
        FROM posts p
        JOIN post_targets t ON t.post_key = p.post_key
        WHERE {" AND ".join(where)}
        ORDER BY p.date_utc DESC, t.target
        """,
        tuple(params),
    ).fetchall()
    return [dict(row) for row in rows]


def apply_backfill_plan(conn, rows, reset_counts: bool = False) -> int:
    now = iso_utc(utc_now_dt())
    for row in rows:
        if reset_counts:
            conn.execute(
                """
                INSERT INTO metric_schedule(post_key, target, next_check_at, check_count, frozen_at, updated_at)
                VALUES (?, ?, NULL, 0, NULL, ?)
                ON CONFLICT(post_key, target) DO UPDATE SET
                    next_check_at=NULL,
                    check_count=0,
                    frozen_at=NULL,
                    last_error=NULL,
                    updated_at=excluded.updated_at
                """,
                (row["post_key"], row["target"], now),
            )
        else:
            conn.execute(
                """
                INSERT INTO metric_schedule(post_key, target, next_check_at, frozen_at, updated_at)
                VALUES (?, ?, NULL, NULL, ?)
                ON CONFLICT(post_key, target) DO UPDATE SET
                    next_check_at=NULL,
                    frozen_at=NULL,
                    last_error=NULL,
                    updated_at=excluded.updated_at
                """,
                (row["post_key"], row["target"], now),
            )
    conn.execute(
        "UPDATE metric_schedule SET frozen_at=?, next_check_at=NULL, updated_at=? WHERE target IN (?, ?)",
        (now, now, *UNSUPPORTED_METRIC_TARGETS),
    )
    conn.commit()
    return len(rows)


def audit_alerts(conn) -> dict:
    rows = conn.execute(
        """
        SELECT severity, event_type, COUNT(*) AS count, MAX(created_at) AS latest
        FROM post_events
        GROUP BY severity, event_type
        ORDER BY severity, event_type
        """
    ).fetchall()
    recent = conn.execute(
        """
        SELECT severity, event_type, target, message, created_at
        FROM post_events
        ORDER BY created_at DESC
        LIMIT 20
        """
    ).fetchall()
    failed_jobs = conn.execute(
        """
        SELECT target, COUNT(*) AS count, MAX(updated_at) AS latest
        FROM publish_jobs
        WHERE status='failed'
        GROUP BY target
        ORDER BY target
        """
    ).fetchall()
    metric_errors = conn.execute(
        """
        SELECT target, COUNT(*) AS count, MAX(updated_at) AS latest
        FROM metric_schedule
        WHERE last_error IS NOT NULL AND last_error != ''
        GROUP BY target
        ORDER BY target
        """
    ).fetchall()
    return {
        "post_events_by_type": [dict(row) for row in rows],
        "recent_post_events": [dict(row) for row in recent],
        "failed_publish_jobs": [dict(row) for row in failed_jobs],
        "metric_schedule_errors": [dict(row) for row in metric_errors],
        "telegram_notifications": "not_audited_in_code",
    }


def active_maintenance_lock(conn) -> dict | None:
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS maintenance_locks (
            name TEXT PRIMARY KEY,
            owner TEXT NOT NULL,
            expires_at TEXT NOT NULL,
            created_at TEXT NOT NULL
        )
        """
    )
    now = datetime.now(timezone.utc).replace(microsecond=0).isoformat()
    row = conn.execute(
        "SELECT name, owner, expires_at, created_at FROM maintenance_locks WHERE name=? AND expires_at >= ?",
        (MAINTENANCE_LOCK_NAME, now),
    ).fetchone()
    return dict(row) if row else None


@contextmanager
def maintenance_lock(conn, timeout_seconds: int = 1):
    now = datetime.now(timezone.utc).replace(microsecond=0)
    expires = (now + timedelta(minutes=30)).isoformat()
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS maintenance_locks (
            name TEXT PRIMARY KEY,
            owner TEXT NOT NULL,
            expires_at TEXT NOT NULL,
            created_at TEXT NOT NULL
        )
        """
    )
    owner = f"{os.uname().nodename}:{os.getpid()}"
    try:
        conn.execute(
            """
            INSERT INTO maintenance_locks(name, owner, expires_at, created_at)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(name) DO UPDATE SET
                owner=CASE WHEN maintenance_locks.expires_at < ? THEN excluded.owner ELSE maintenance_locks.owner END,
                expires_at=CASE WHEN maintenance_locks.expires_at < ? THEN excluded.expires_at ELSE maintenance_locks.expires_at END,
                created_at=CASE WHEN maintenance_locks.expires_at < ? THEN excluded.created_at ELSE maintenance_locks.created_at END
            """,
            (MAINTENANCE_LOCK_NAME, owner, expires, now.isoformat(), now.isoformat(), now.isoformat(), now.isoformat()),
        )
        row = conn.execute("SELECT owner FROM maintenance_locks WHERE name=?", (MAINTENANCE_LOCK_NAME,)).fetchone()
        if row and row["owner"] != owner:
            raise RuntimeError(f"maintenance lock is held by {row['owner']}")
        conn.commit()
        yield
    finally:
        conn.execute("DELETE FROM maintenance_locks WHERE name=? AND owner=?", (MAINTENANCE_LOCK_NAME, owner))
        conn.commit()


def backup_database(db_path: Path, backup_dir: Path | None = None) -> Path:
    backup_dir = backup_dir or db_path.parent / "backups"
    backup_dir.mkdir(parents=True, exist_ok=True)
    stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    backup_path = backup_dir / f"{db_path.stem}-{stamp}.db"
    src = sqlite3.connect(str(db_path))
    try:
        dest = sqlite3.connect(str(backup_path))
        try:
            src.backup(dest)
        finally:
            dest.close()
    finally:
        src.close()
    return backup_path

File: apps/backend/src/ops/maintenance.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { and, eq, lt } from "drizzle-orm";
import type { BackendDb } from "../db/client.js";
import { maintenanceLocks } from "../db/schema.js";

export async function backupDatabase(backendDb: BackendDb, sourcePath: string, destinationDirectory?: string): Promise<string> {
  if (sourcePath === ":memory:") throw new Error("cannot back up an in-memory database");
  const directory = destinationDirectory ?? path.join(path.dirname(sourcePath), "backups");
  fs.mkdirSync(directory, { recursive: true });
  const stamp = new Date()
    .toISOString()
    .replace(/[-:]/g, "")
    .replace(/\.\d{3}Z$/, "Z");
  const destination = path.join(directory, `${path.basename(sourcePath, path.extname(sourcePath))}-${stamp}.db`);
  await backendDb.sqlite.backup(destination);
  backendDb.sqlite
    .prepare("INSERT INTO deployment_snapshots(action,status,backup_path,created_at) VALUES ('backup','ok',?,?)")
    .run(destination, new Date().toISOString());
  return destination;
}

export function restoreDatabase(source: string, destination: string, force: boolean): void {
  if (!force) throw new Error("restore requires --force");
  if (!fs.existsSync(source)) throw new Error(`backup does not exist: ${source}`);
  fs.mkdirSync(path.dirname(destination), { recursive: true });
  for (const suffix of ["", "-wal", "-shm"]) fs.rmSync(`${destination}${suffix}`, { force: true });
  fs.copyFileSync(source, destination, fs.constants.COPYFILE_FICLONE);
}

export function buildMetricsBackfillPlan(
  backendDb: BackendDb,
  options: { targets: string[]; refs?: string[]; dateFrom?: string; dateTo?: string },
): Record<string, unknown>[] {
  if (options.targets.length === 0) return [];
  const where = ["p.status='active'", "t.status='published'", `t.target IN (${options.targets.map(() => "?").join(",")})`];
  const parameters: unknown[] = [...options.targets];
  if (options.refs?.length) {
    where.push(`p.post_key IN (${options.refs.map(() => "?").join(",")})`);
    parameters.push(...options.refs);
  }
  if (options.dateFrom) {
    where.push("p.date_utc >= ?");
    parameters.push(options.dateFrom);
  }
  if (options.dateTo) {
    where.push("p.date_utc <= ?");
    parameters.push(options.dateTo);
  }
  return backendDb.sqlite
    .prepare(
      `SELECT p.post_key,p.post_id,p.message_id,p.date_utc,t.target FROM posts p JOIN post_targets t ON t.post_key=p.post_key WHERE ${where.join(" AND ")} ORDER BY p.date_utc DESC,t.target`,
    )
    .all(...parameters) as Record<string, unknown>[];
}

export function applyMetricsBackfill(backendDb: BackendDb, rows: Record<string, unknown>[], resetCounts = false): number {
  const now = new Date().toISOString();
  backendDb.sqlite.transaction(() => {
    for (const row of rows) {
      backendDb.sqlite
        .prepare(`INSERT INTO metric_schedule(post_key,target,next_check_at,check_count,frozen_at,last_error,updated_at)
        VALUES (?,?,NULL,0,NULL,NULL,?) ON CONFLICT(post_key,target) DO UPDATE SET next_check_at=NULL,
        check_count=${resetCounts ? "0" : "metric_schedule.check_count"},frozen_at=NULL,last_error=NULL,updated_at=excluded.updated_at`)
        .run(row.post_key, row.target, now);
    }
    backendDb.sqlite
      .prepare("UPDATE metric_schedule SET frozen_at=?,next_check_at=NULL,updated_at=? WHERE target IN ('x','linkedin')")
      .run(now, now);
  })();
  return rows.length;
}

export function auditOperations(backendDb: BackendDb): Record<string, unknown> {
  return {
    postEventsByType: backendDb.sqlite
      .prepare(
        "SELECT severity,event_type,count(*) AS count,max(created_at) AS latest FROM post_events GROUP BY severity,event_type ORDER BY severity,event_type",
      )
      .all(),
    recentPostEvents: backendDb.sqlite
      .prepare("SELECT severity,event_type,target,message,created_at FROM post_events ORDER BY created_at DESC LIMIT 20")
      .all(),
    failedPublishJobs: backendDb.sqlite
      .prepare(
        "SELECT target,count(*) AS count,max(updated_at) AS latest FROM publish_jobs WHERE status='failed' GROUP BY target ORDER BY target",
      )
      .all(),
    metricScheduleErrors: backendDb.sqlite
      .prepare(
        "SELECT target,count(*) AS count,max(updated_at) AS latest FROM metric_schedule WHERE last_error IS NOT NULL AND last_error != '' GROUP BY target ORDER BY target",
      )
      .all(),
  };
}

export function withMaintenanceLock<T>(backendDb: BackendDb, operation: () => T): T {
  const name = "metrics_maintenance";
  const owner = `${os.hostname()}:${process.pid}`;
  const now = new Date();
  const expires = new Date(now.getTime() + 30 * 60_000).toISOString();
  backendDb.db.transaction((tx) => {
    tx.delete(maintenanceLocks)
      .where(and(eq(maintenanceLocks.name, name), lt(maintenanceLocks.expiresAt, now.toISOString())))
      .run();
    tx.insert(maintenanceLocks).values({ name, owner, expiresAt: expires, createdAt: now.toISOString() }).onConflictDoNothing().run();
    const row = tx.select({ owner: maintenanceLocks.owner }).from(maintenanceLocks).where(eq(maintenanceLocks.name, name)).get();
    if (!row) throw new Error("maintenance lock could not be acquired");
    if (row.owner !== owner) throw new Error(`maintenance lock is held by ${row.owner}`);
  });
  try {
    return operation();
  } finally {
    backendDb.db
      .delete(maintenanceLocks)
      .where(and(eq(maintenanceLocks.name, name), eq(maintenanceLocks.owner, owner)))
      .run();
  }
}
